README.md
# Mini-Circuits Python Interfaces

This package provides interfaces to various Mini Circuits USB devices.

### Device Support

| Part            | Type                  | Status                          |
|-----------------|-----------------------|---------------------------------|
| RCM series      | mechanical switches   | Not supported yet               |
| ZTM series      | mechanical switches   | Not supported yet               |
| RC series       | mechanical switches   | Not supported yet               |
| USB series      | mechanical switches   | Not supported yet               |
| U2C-1SP2T-63VH  | solid-state switch    | Not supported yet               |
| U2C-1SP4T-63H   | solid-state switch    | Not supported yet               |
| USB-1SP16T-83H  | solid-state switch    | Not supported yet               |
| USB-1SP8T-63H   | solid-state switch    | Not supported yet               |
| USB-2SP2T-DCH   | solid-state switch    | Not supported yet               |
| USB-4SP2T-63H   | solid-state switch    | Not supported yet               |
| USB-SP4T-63     | solid-state switch    | Verified working :+1:           |
| PWR-2.4GHS-75   | power sensor          | Might work?                     |
| PWR-4GHS        | power sensor          | Might work?                     |
| PWR-4RMS        | power sensor          | Might work?                     |
| PWR-6GHS        | power sensor          | Verified working :+1:           |
| PWR-6LGHS       | power sensor          | Might work?                     |
| PWR-6LRMS-RC    | power sensor          | Might work?                     |
| PWR-6RMS-RC     | power sensor          | Might work?                     |
| PWR-8P-RC       | power sensor          | Might work?                     |
| PWR-8FS         | power sensor          | Might work?                     |
| PWR-8GHS        | power sensor          | Might work?                     |
| PWR-8GHS-RC     | power sensor          | Might work?                     |
| RC4DAT-6G-60    | 4-channel attenuator  | Not supported yet               |
| RC4DAT-6G-95    | 4-channel attenuator  | Not supported yet               |
| RCDAT-30G-30    | attenuator            | Might work via USB?             |
| RCDAT-3000-63W2 | attenuator            | Might work via USB?             |
| RCDAT-4000-120  | attenuator            | Might work via USB?             |
| RCDAT-6000-30   | attenuator            | Might work via USB?             |
| RCDAT-6000-60   | attenuator            | Might work via USB?             |
| RCDAT-6000-90   | attenuator            | Might work via USB?             |
| RCDAT-6000-110  | attenuator            | Verified working via USB :+1:   |
| RCDAT-8000-30   | attenuator            | Might work via USB?             |
| RCDAT-8000-60   | attenuator            | Might work via USB?             |
| RCDAT-8000-90   | attenuator            | Might work via USB?             |
| RUDAT-13G-90    | attenuator            | Might work?                     |
| RUDAT-13G-60    | attenuator            | Might work?                     |
| RUDAT-4000-120  | attenuator            | Might work?                     |
| RUDAT-6000-30   | attenuator            | Might work?                     |
| RUDAT-6000-60   | attenuator            | Verified working :+1:           |
| RUDAT-6000-90   | attenuator            | Might work?                     |
| RUDAT-6000-110  | attenuator            | Might work?                     |
| ZVVA-3000       | attenuator            | Might work?                     |
| UFC-6000        | frequency counter     | Not supported yet               |
| FCPM-6000RC     | frequency and power   | Not supported yet               |
| SSG-6000RC      | signal generator      | Not supported yet               |
| SSG-6001RC      | signal generator      | Not supported yet               |


### Quick Start

Targeted to Python 3, works on Python 2 as well.

Install dependencies (Ubuntu):

    $ sudo apt install libudev-dev

Install the package:

    $ pip install pyminicircuits

Create a new udev rules file in ``/etc/udev/rules.d/85-minicircuits.rules``
with:

    ATTR{idVendor}=="20ce", MODE="660", GROUP="plugdev"

This will tell udev to allow members of the plugdev group to access all
Mini-Circuits USB devices.

Ensure that your user is a member of the plugdev group:

    $ sudo usermod -a -G plugdev <username>

Reload the udev rules:

    $ sudo udevadm control --reload-rules
    $ sudo udevadm trigger

Connect to a power sensor and start measuring power:

    $ minicircuits-powersensor

### API Example

Connect to an RUDAT attenuator and set values:

    from pyminicircuits import Attenuator

    a = Attenuator()
    print(a.get_serial())
    a.set_attenuation(32.5)
    print(a.get_attenuation())

#### License

Pyminicircuits is licensed under an MIT license. Please see the LICENSE file
for more information.


CHANGES
Pyminicircuits Changelog
========================

Version 0.0.1
-------------

- Initial release.


pyminicircuits/cmd/powersensor.py
import sys
import time
import argparse

from .. import PowerSensor


def main(argv=sys.argv):
    p = argparse.ArgumentParser(
        description='Interact with Mini Circuits USB power sensors.')

    p.add_argument('--verbose', action='store_true')
    p.add_argument('--mode', default='low-noise')
    p.add_argument('--freq', type=float, default=10e6)
    p.add_argument('--interval', type=float, default=0.5)
    opts = p.parse_args(argv[1:])

    sensor = PowerSensor()

    if opts.verbose:
        print("Model name: %s" % sensor.get_model_name())
        print("Serial: %s" % sensor.get_serial())
        print("Firmware version: %s" % sensor.get_firmware_version())
        print("Setting measurement mode...")

    sensor.set_measurement_mode(opts.mode)

    if opts.verbose:
        print("Target signal frequency: %s" % opts.freq)

    while True:
        temp = sensor.get_temperature()
        power = sensor.get_power(opts.freq)
        print("%s C / %s dBm" % (temp, power))
        time.sleep(opts.interval)


pyminicircuits/__init__.py
import time
import hid


INSTALL_STEPS = """Failed to open device. You may need to follow these steps:

Install dependencies:

    $ sudo apt install libudev-dev
    $ pip install hidapi

Create a new file /etc/udev/rules.d/85-minicircuits.rules with:

    ATTR{idVendor}=="20ce", MODE="660", GROUP="plugdev"

This will tell udev to allow members of the plugdev group to access all
Mini-Circuits USB devices.

Ensure that your user is a member of the plugdev group:

    $ sudo usermod -a -G plugdev <username>

Reload the udev rules:

    $ sudo udevadm control --reload-rules
    $ sudo udevadm trigger
"""


class BaseInterface(object):
    DEFAULT_VID = 0x20ce

    def __init__(self, vid=None, pid=None):
        vid = vid or self.DEFAULT_VID
        pid = pid or self.DEFAULT_PID
        self.h = hid.device()
        try:
            self.h.open(vid, pid)
        except OSError:
            print(INSTALL_STEPS)
            raise
        self.h.set_nonblocking(1)

    @classmethod
    def parse_response_string(cls, data):
        chars = []
        for b in data[1:]:
            if b == 0:
                break
            chars.append(chr(b))
        return ''.join(chars)

    def _cmd(self, *data):
        cmd = [0] * 64
        for index, data in enumerate(data):
            cmd[index] = data
        self.h.write(cmd)
        time.sleep(0.05)
        while True:
            d = self.h.read(64)
            if d:
                break
        if d[0] != cmd[0]:
            raise RuntimeError("Invalid response from device: %s" % d)
        return d


class PowerSensor(BaseInterface):
    """
    Class for interfacing with Mini-Circuits USB Power Sensors.

    Verified working:
        - PWR-6GHS
    """

    # Mini-Circuits PWR-6GHS USB Power Sensor
    DEFAULT_PID = 0x11

    # Not sure if any of these command codes are also shared.
    CMD_GET_MODEL_NAME = 104
    CMD_GET_SERIAL = 105
    CMD_SET_MEASUREMENT_MODE = 15
    CMD_READ_POWER = 102
    CMD_GET_TEMPERATURE = 103
    CMD_GET_FIRMWARE_VERSION = 99

    def get_model_name(self):
        d = self._cmd(self.CMD_GET_MODEL_NAME)
        return self.parse_response_string(d)

    def get_serial(self):
        d = self._cmd(self.CMD_GET_SERIAL)
        return self.parse_response_string(d)

    def set_measurement_mode(self, mode='low-noise'):
        mode_num = {
            'low-noise': 0,
            'fast-sampling': 1,
            'fastest-sampling': 2,
        }[mode]
        self._cmd(self.CMD_SET_MEASUREMENT_MODE, mode_num)

    def get_power(self, freq):
        if freq > 65.535e6:
            scale = 1e6
            units = 'M'
        else:
            scale = 1e3
            units = 'k'
        freq = int(round(freq / scale))
        freq1 = freq >> 8
        freq2 = freq - (freq1 << 8)
        d = self._cmd(self.CMD_READ_POWER, freq1, freq2, ord(units))
        s = ''.join(chr(c) for c in d[1:7])
        return float(s)

    def get_temperature(self):
        d = self._cmd(self.CMD_GET_TEMPERATURE)
        s = ''.join(chr(c) for c in d[1:7])
        return float(s)

    def get_firmware_version(self):
        d = self._cmd(self.CMD_GET_FIRMWARE_VERSION)
        return chr(d[5]) + chr(d[6])


class SwitchAttenuatorBase(BaseInterface):
    CMD_GET_PART_NUMBER = 40
    CMD_GET_SERIAL_NUMBER = 41

    def get_part_number(self):
        d = self._cmd(self.CMD_GET_PART_NUMBER)
        return self.parse_response_string(d)

    def get_serial(self):
        d = self._cmd(self.CMD_GET_SERIAL_NUMBER)
        return self.parse_response_string(d)


class Attenuator(SwitchAttenuatorBase):
    """
    Class for interfacing with Mini-Circuits USB attenuators.

    Verified working:
        - RCDAT-6000-110
        - RUDAT-6000-60
    """
    # Mini-Circuits RUDAT-6000-60
    DEFAULT_PID = 0x23

    CMD_GET_ATTENUATION = 18
    CMD_SET_ATTENUATION = 19

    def get_attenuation(self):
        d = self._cmd(self.CMD_GET_ATTENUATION)
        full_part = d[1]
        frac_part = float(d[2]) / 4.0
        return full_part + frac_part

    def set_attenuation(self, value):
        value1 = int(value)
        value2 = int((value - value1) * 4.0)
        self._cmd(self.CMD_SET_ATTENUATION, value1, value2)


class Switch(SwitchAttenuatorBase):
    """
    Class for interfacing with Mini-Circuits USB switches.

    Verified working:
        - USB-SP4T-63
    """
    # Mini-Circuits USB-SP4T-63
    DEFAULT_PID = 0x22

    CMD_GET_SWITCH_PORT = 15

    def set_active_port(self, port):
        """
        Set which port is connected to the COM port: 1-indexed.
        """
        if port not in (1, 2, 3, 4):
            raise ValueError("Invalid switch port: %s" % port)
        self._cmd(port)

    def get_active_port(self):
        """
        Return which port is connected to the COM port: 1-indexed.
        """
        d = self._cmd(self.CMD_GET_SWITCH_PORT)
        port = d[1]
        return port
